Document autoLogin configuration #171 (#172)
enonic/app-oidc-idprovider: docs/index.adoc

You will need to gather the following information from the Authorization Server setup, for the <<#id_provider_creation, ID Provider configuration>>

> . The Client ID. Must be set, if log in via form is enabled.

Reviewer: ```suggestion
. The Client ID. Not required authentication is done only by Auto Login.
```

Author: done

> . The Client secret. Must be set, if log in via form is enabled.

Reviewer: ```suggestion
. The Client secret. Not required authentication is done only by Auto Login.
```

. The Issuer Identifier (URL)
. The Authorization endpoint URL
. The Token endpoint URL
. The JWKS URI

> . The `oidcWellKnownEndpoint` (OpenID Provider Configuration URL), to obtain the values for the above fields and for `userinfoUrl`.

Reviewer: ```suggestion
. The `oidcWellKnownEndpoint` (Provider's OpenID Connect Discovery URL), to obtain the values for the above fields and for `userinfoUrl`.
```

You might have been given only the Issuer URL instead of the list of endpoint URLs.
In that case the endpoint URLs are accessible under: `[Issuer URL]/.well-known/openid-configuration`

> The JSON result will contain the required information: `issuer`, `authorization_endpoint`, `token_endpoint` and `jwks_uri`.

Reviewer: ```suggestion
The JSON result will contain the additional information: `issuer`, `authorization_endpoint`, `token_endpoint` and `jwks_uri`.
```

idprovider.<idprovidername>.endSession.additionalParameters.0.key=  (string, required)
idprovider.<idprovidername>.endSession.additionalParameters.0.value=  (string, required)

idprovider.<idprovidername>.clientId=  (string, optional)
idprovider.<idprovidername>.clientSecret=  (string, optional)

> idprovider.<idprovidername>.mappings.displayName=  (string, required, default to ${userinfo.preferred_username})

Reviewer: ```suggestion
idprovider.<idprovidername>.mappings.displayName=  (string, required, defaults to ${userinfo.preferred_username})
```

> idprovider.<idprovidername>.mappings.email=  (string, required, default to ${userinfo.email})

Reviewer: ```suggestion
idprovider.<idprovidername>.mappings.email=  (string, required, defaults to ${userinfo.email})
```

==== Fields in the config

* `autoinit`: Automatic initialization. If the config file contains `autoinit=true`, then during startup this app will automatically create ID providers for all settings declared in the file, if they don't already exist. For example, `idprovider.myfirstidp.someKey=someValue` and `idprovider.anotheridp.anotherKey=anotherValue` will declare two idproviders named `myfirstidp` and `anotheridp`.

> * `displayName`: Display name of ID Provider in User app.

Reviewer: ```suggestion
* `displayName`: Display name of ID Provider in Users app.
```

> * `description`: Description of ID Provider in User app.

Reviewer: ```suggestion
* `description`: Description of ID Provider in Users app.
```

  ** `email`: Template for the email in format of `@@{expression}`, e.g. `@@{userinfo.email}`.
* `defaultGroups` (`idprovider.<idprovidername>.defaultGroups`): Groups to assign to this user on creation. A list of space-separated group keys. The key must be in the format `group:[idprovidername]:[groupname]`, e.g. `group:myoidcprovider:authors`.

**Auto Login**

* `autoLogin...` (`idprovider.<idprovidername>.autoLogin...` namespace): Auto login configuration. To enable auto login, the `jwksUri` config must be set.
  ** `createUser`: Create a user. If set to `true`, a user will be created automatically if they don't exist.
  ** `createSession`: Create a session. If set to `true`, a session will be established upon the user's login. By default, the user will be logged in with `REQUEST` scope.
  ** `wsHeader`: Look for a token in the Sec-WebSocket-Protocol header. If set to `true`, the app will look for a token in the Sec-WebSocket-Protocol header.
  ** `allowedAudience`: Allowed audience. A list of space-separated strings. If set, the app will only accept tokens with an audience that matches one of the values in this list.

> To disable the ability to log in via the form, `clientId` and `clientSecret` must be omitted.

Reviewer: ```suggestion
To disable the ability to auth by Authorization Code Flow, `clientId` should be omitted.
```

{zwsp} +

== Upgrade Notes

=== Version 3

Version `3.0` of this app introduces a breaking change:

- To verify the token signature the `jwksUri` field is now required in the ID Provider configuration.

> - The `clientId` and `clientSecret` fields are now optional in the ID Provider configuration. If you want to disable the ability to log in via the form, these fields must be omitted.

Reviewer: ```suggestion
- The `clientId` and `clientSecret` fields are now optional in the ID Provider configuration. If you want to disable the ability to auth by Authorization Code Flow, `clientId` should be omitted.
```
